gov/politeia/util.go

```diff
@@ -0,0 +1,154 @@
+package politeia
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/decred/politeia/politeiad/plugins/usermd"
+	piv1 "github.com/decred/politeia/politeiawww/api/pi/v1"
+	recordsv1 "github.com/decred/politeia/politeiawww/api/records/v1"
+	ticketvotev1 "github.com/decred/politeia/politeiawww/api/ticketvote/v1"
+)
+
+// userMetadataDecode returns the parsed data for the usermd plugin metadata
+// stream.
+func userMetadataDecode(ms []recordsv1.MetadataStream) (*usermd.UserMetadata, error) {
+	var userMD *usermd.UserMetadata
+	for _, m := range ms {
+		if m.PluginID != usermd.PluginID ||
+			m.StreamID != usermd.StreamIDUserMetadata {
+			// This is not user metadata
+			continue
+		}
+		var um usermd.UserMetadata
+		err := json.Unmarshal([]byte(m.Payload), &um)
+		if err != nil {
+			return nil, err
+		}
+		userMD = &um
+		break
+	}
+	return userMD, nil
+}
+
+// proposalMetadataDecode returns the parsed data for the pi plugin proposal
+// metadata stream.
+func proposalMetadataDecode(fs []recordsv1.File) (*piv1.ProposalMetadata, error) {
+	var pmp *piv1.ProposalMetadata
+	for _, f := range fs {
+		if f.Name != piv1.FileNameProposalMetadata {
+			continue
+		}
+		b, err := base64.StdEncoding.DecodeString(f.Payload)
+		if err != nil {
+			return nil, err
+		}
+		var pm piv1.ProposalMetadata
+		err = json.Unmarshal(b, &pm)
+		if err != nil {
+			return nil, err
+		}
+		pmp = &pm
+		break
+	}
+	if pmp == nil {
+		return nil, fmt.Errorf("proposal metadata not found")
+	}
+	return pmp, nil
+}
+
+// statusTimestamps contains the published, censored and abandoned timestamps
+// that are used on dcrdata's UI.
+type statusTimestamps struct {
+	published int64
+	censored  int64
+	abandoned int64
+}
+
+// statusChangeMetadataDecode returns the published, censored and abandoned
+// dates from status change metadata streams in the statusTimestamps struct.
+// It also returns the status change message for the latest metadata stream.
+func statusChangeMetadataDecode(md []recordsv1.MetadataStream) (*statusTimestamps, string, error) {
+	statuses := make([]usermd.StatusChangeMetadata, 0, 16)
```

Comment:
That's just a random number politeia uses that's big enough for the status changes. As of now, a record can only undergo a status change two times at max, but this might change in the future.

reference: [link](https://github.com/decred/politeia/blob/84a52cfc0975cb1ca2df676d15e7ad0637a8ce42/politeiad/backendv2/tstorebe/plugins/usermd/hooks.go#L267)

Reply:
Huh, alright.  Normally I'd request a comment and/or const for a magic number like this, esp when it's not even necessary to prealloc at all.  For just a handful of status changes, where the number is unknown, I'd go with `var statuses []usermd.StatusChangeMetadata`.

Reply:
Makes sense mate, sounds good